pkg/controller/workload/job/job_controller.go

```diff
@@ -619,14 +660,12 @@ func jobAndWorkloadEqual(job *batchv1.Job, wl *kueue.Workload) bool {
 		return false
 	}
 
-	// nodeSelector may change, hence we are not checking for
-	// equality of the whole job.Spec.Template.Spec.
-	if !equality.Semantic.DeepEqual(job.Spec.Template.Spec.InitContainers,
-		wl.Spec.PodSets[0].Spec.InitContainers) {
+	specHash, err := jobSpecHash(job)
```

Comment:
I think this is not what we want, before we validate that the job is semantically equal with workload, for workload might be created randomly, but now we just validate the job is not updated, if I understood correctly, then, we can compare the UID and ResourceVersion instead.

Reply:
The difference between  `jobSpecHash` and UID+Resourece is taht the result of `jobSpecHash` will be the same ti the [Init]Containers specs are not changed.

From the equality point of view, indeed, now we check that the current version of the job is semantically equal to the job version at the creation of the workload, however , starting from the semantically equal job and applying the same default logic will create the same workload (the only exception for this being if LimitRanges change ).

Reply:
this question is still open

Reply:
The only scenario in which could change the the the flow is when a workload is not scheduled because the an old default value douse not fit and it will fit with the new one.
What we could do is to extend the workload controller to watch the changes in the  LimitsRanges and mark the workloads using outdated values.  Then the job controllers can decide if they want to delete and recreate the workload. Do we want something like that?

Reply:
Let's assume LimitRanges never change. In that case, this is a fine solution and I would be ok merging it.

Realistically, LimitRanges can change, the next question is whether it's acceptable to be out-of-sync for a first iteration of LimitRange support. I think it's fine.

However, I'm afraid that once we add support for changing LimitRanges, we would just throw away this code based on hashes. Are we ok with that?

I think it's just more useful to directly write the code that handles changing LimitRanges. We don't need to add code to react to changes in LimitRange. It suffices to do the checks whenever there are changes in the Workload or Job object. If there are changes, we can just keep the current logic that deletes the Workload object. Once we decide to have a better support for this, we can change the code to update the Workload object instead.

Reply:
This part just keeps track of the sync between the job spec and the original (before defaulting is applied ) workload content and should not be impacted by the the limits changes.

Also #611 was opened to work on that resync.